### PR TITLE
[ONNX] Track and list model params for scripting

### PR DESCRIPTION
**Pull request by neginraoof**
List model parameters as inputs following freezing script module.

### CODE REVIEW

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+  Node* m = *block->nodes().begin();
+  WithInsertPoint guard(m);
+
+  while (!blocks.empty()) {
+    Block* block = blocks.back();
+    blocks.pop_back();
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* n = *it;
+      it++; // node n can be destroyed
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:74]: If node `n` is destroyed, does it affect `block`? If so, does `it++` work properly after `n` gets destroyed?

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+      for (Block* sub_block : n->blocks()) {
+        blocks.emplace_back(sub_block);
+      }
+      if (n->kind() == prim::SetAttr) {
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:79]: What is difference between `prim::SetAttr` and `prim::GetAttr` types?

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+        findSubModuleAttr(input, name, attrModule, graph);
+
+        TORCH_INTERNAL_ASSERT(attrModule.hasattr(name));
+        Value* paramConst = nullptr;
+
+        auto attr = attrModule.attr(name);
+
+        std::string fullName("self_");
+        for (auto& name : names_) {
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:96]: We have `name` out of this loop scope but in this function, then it is a little confusing to read, shall we use other variable name instead?

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp (hunk 2 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:79]: SetAttr writes to an attribute.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp (hunk 1 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:74]: If n is destroyed (even if block is emptied), it won't affect the block. it++ always point to the next node so that destroying n won't be a problem.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* n = *it;
+      it++; // node n can be destroyed
+
+      for (Block* sub_block : n->blocks()) {
+        blocks.emplace_back(sub_block);
+      }
+      if (n->kind() == prim::SetAttr &&
+          n->s(attr::name) == "num_batches_tracked") {
+        n->destroy();
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:83]: This is special case handling for batch norm right? Feels like this should move to the part of code getting rid of GetAttr/SetAttr.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+//   ...
+std::deque<std::string> findSubModuleAttr(
+    Value* input,
+    std::string& name,
+    Module& attrModule,
+    std::shared_ptr<Graph>& graph) {
+  Node* node = input->node();
+  std::deque<std::string> moduleNames;
+
+  while (!(node->outputs()[0]->type() == graph->inputs()[0]->type())) {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:29]:
```suggestion
  while (node->outputs().at(0)->type() != graph->inputs().at(0)->type()) {
```
`at` is safer as it checks out of bounds

Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:29]: Could you explain what this condition is guarding against?

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    if (node->kind() == prim::GetAttr) {
+      moduleNames.push_front(node->s(attr::name));
+      node = node->inputs()[0]->node();
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:32]: nit: same

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+          fullName += name + '_';
+        }
+        fullName += name;
+
+        auto type = attrModule.type();
+        auto slot = *type->findAttributeSlot(name);
+
+        if (type->is_parameter(slot) || type->is_buffer(slot) ||
+            name == "training") {
+          if (type->is_parameter(slot) || type->is_buffer(slot) ||
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:109]: what's the difference between this and line 107? ...

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+              name == "training") {
```
Comment (BowenBao) [inline torch/csrc/jit/passes/onnx/list_model_parameters.cpp:110]: why is `training` a factor?

Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:110]: Training is an attribute for torch script models, providing a dynamic mode for training/eval.
Though in ONNX, we don't want to keep it as a parameter as models are exported either in eval or training mode. So keep this as a constant.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp (hunk 7 above)
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:109]: Duplicate, thanks!

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+    attrModule = attrModule.attr(moduleName).toModule();
+  }
+  return moduleNames;
+}
+
+Value* addParamAsArgument(Function* function, std::string& name, IValue& attr) {
+  auto schema = function->getSchema();
+  auto args = schema.arguments();
+  args.emplace_back(Argument(name, nullptr, c10::nullopt, attr));
+  auto new_schema = FunctionSchema(
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:52]: This is similar to logic in freeze_module pass. But I can check this.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+  std::deque<std::string> moduleNames;
+
+  // Loop starts from inner submodule and follows the chain until reaches the
+  // top module.
+  while (node->outputs().at(0)->type() != graph->inputs().at(0)->type()) {
+    if (node->kind() == prim::GetAttr) {
+      moduleNames.push_front(node->s(attr::name));
+      node = node->inputs()[0]->node();
+    } else {
+      return moduleNames;
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:37]: Yeah this makes sense, thanks!

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            auto tensor_ = attr.toTensor();
+            if (isEval && tensor_.requires_grad()) {
+              tensor_ = tensor_.detach();
+              tensor_.set_requires_grad(false);
+              attr = IValue(tensor_);
+            }
+            attrValues.emplace_back(attr.toTensor());
+            paramConst = addParamAsArgument(function_, fullName, attr);
+          } else if (
+              attr.isObject() && !attr.toObjectRef().type()->is_module()) {
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:130]: Yes, this is actually currently supporting quantization class types provided here. Will need to follow up with Elias on support for other defined classes.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+        if (!attrModule.hasattr(name)) {
+          continue;
+        }
+        Value* paramConst = nullptr;
+
+        auto attr = attrModule.attr(name);
+
+        std::string fullName("self_");
+        for (auto& name : moduleNames) {
+          fullName += name + '_';
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:109]: So named_parameters() and named_buffers() do not have the "self." prefix as part of the name.
So param name can be: %self.conv.weight for a named_parameter conv.weight.
Do you think to remove "self." from param names?

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:130]: For each class, the __getstate__ method returns the attribute that is then tracked in model params. This only works for tensor attributes.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+std::pair<Module, std::vector<IValue>> list_module_parameters(
+    const Module& module) {
+  Module moduleClone = module.clone(true);
+  Method method = moduleClone.get_method("forward");
+  auto function = &method.function();
+  std::vector<IValue> modelParams;
+
+  GRAPH_DEBUG("List attributes for function: " + function->name());
+  auto graph = function->graph();
+  auto attributes = getParamAttributes(graph, moduleClone, function);
```
[the remark this answers is not in the document]
Comment (neginraoof) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp:159]: Sure. But to clarify, we are tracking both named_parameters and named_buffers.

File: torch/csrc/jit/passes/onnx/list_model_parameters.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (BowenBao) [reply torch/csrc/jit/passes/onnx/list_model_parameters.cpp]: Yes, as long as they match with named_parameters() and named_buffers().

**Review by BowenBao — APPROVED** [review]
Looks good, please ensure that the parameter names match with `named_parameters()` and `named_buffers()`, preferably add a test to cover it.

**Review by bzinodev — APPROVED** [review]

**Comment by agolynski** [comment]
This PR could be causing failures
"Unexpected keyword argument "preserveParameters" for "freeze_module"

Is it possible that you forgot to update torch/_C/__init__.pyi.in to include that param in _freeze_module interface?

**Comment by agolynski** [comment]
Apologies, this is not fault of this PR, but the issue is with https://github.com/pytorch/pytorch/pull/48782 which added outdated mypy annotations.
